### ISSUE TITLE
`mul!` in LinearLayer is calling a "slow" function

### ISSUE THREAD

**Issue by CheukHinHoJerry** [ACEsuit/Polynomials4ML.jl#62]
Currently this is used in `LinearLayer`:

```julia
mul!(unwrap(out), unwrap(x), transpose(ps.W)); 
```
which `mul!` seems to be calling `generic_matmul` because `tranpose(ps.W)` is a `transpose`, which is slow. (Actually in the previous benchmark I didn't observe this for some reason or I just missed it?)

So there are two ways that I can think of:

This wrap the weight matrix `ps.W` as a `PtrArray`, and then transpose returns a `PtrArray` and then we make sure we are calling `matmul!` from `BLAS`

```julia
mul!(unwrap(out), unwrap(X), transpose(PtrArray(ps.W)))
```

Or we hope the input `X` is a `CachedArray/TmpArray/PtrArray`, and then write to the transpose of the `PtrArray` of `out` so that `out` return the correct thing:

```julia
mul!(transpose(unwrap(out)), ps.W, transpose(unwrap(X)))
```

**Comment by cortner** [comment]
thanks for spotting this. It is odd. BLAS has operations for transposed matrices - so why aren't they used? 

Do matrix operations on `PtrArrays` not use BLAS, but one of the new pure Julia implementations? 

My inclination is to try this first: 
```julia
mul!(unwrap(out), unwrap(X), transpose(PtrArray(ps.W)))
```

**Comment by CheukHinHoJerry** [comment]
This looks like something that we can consider using:

https://github.com/JuliaLinearAlgebra/ArrayLayouts.jl

**Comment by CheukHinHoJerry** [comment]
We might not need this now but it looks useful to me, another package by Sheehan Olver:

https://github.com/JuliaLinearAlgebra/ArrayLayouts.jl

**Comment by CheukHinHoJerry** [comment]
Some quick tests:

```julia
using Polynomials4ML
using LuxCore
using StrideArrays
using ObjectPools
using BenchmarkTools
using Random

P4ML = Polynomials4ML

in_d, out_d = 4, 3 # feature dimensions
N = 10 # batch size

# set up
l = P4ML.LinearLayer(in_d, out_d; feature_first = false)
ps, st = LuxCore.setup(MersenneTwister(1234), l)

# assume the input is a PtrArray
X = randn(N, in_d)
X = PtrArray(X)

# acquire for inplace mul!
out = acquire!(st.pool, :bA, (size(X, 1), l.out_dim), eltype(X)); 

# current
@btime mul!(unwrap($out), unwrap($X), transpose($ps.W))

# new
@btime mul!(unwrap($out), unwrap($X), transpose(PtrArray($ps.W)))
@btime mul!(transpose(unwrap($out)), $ps.W, transpose(unwrap($X)))
```
```
using Polynomials4ML
using LuxCore
using StrideArrays
using ObjectPools
using BenchmarkTools
using Random

P4ML = Polynomials4ML

in_d, out_d = 4, 3 # feature dimensions
N = 10 # batch size

# set up
l = P4ML.LinearLayer(in_d, out_d; feature_first = false)
ps, st = LuxCore.setup(MersenneTwister(1234), l)

# assume the input is a PtrArray
X = randn(N, in_d)
X = PtrArray(X)

# acquire for inplace mul!
out = acquire!(st.pool, :bA, (size(X, 1), l.out_dim), eltype(X)); 

# current
@btime mul!(unwrap($out), unwrap($X), transpose($ps.W)) # 658.390 ns, generic matmul, 2 allocations

# new
@btime mul!(unwrap($out), unwrap($X), transpose(PtrArray($ps.W))) # 26.926 ns, BLAS, 0 allocations
@btime mul!(transpose(unwrap($out)), $ps.W, transpose(unwrap($X))) # 27.699 ns, BLAS, 0 allocations
```

with larger matrix size:

```
 694.274 μs (3 allocations: 30.75 KiB)
  34.621 μs (0 allocations: 0 bytes)
  34.642 μs (0 allocations: 0 bytes)
```

Not sure where the extra allocation comes from but that's not important. I guess I inclined to 

```julia
mul!(unwrap(out), unwrap(X), transpose(PtrArray(ps.W)))
```

too. I will create a PR to fix it later.

**Comment by cortner** [comment]
thanks for the careful testing.

**Comment by cortner** [comment]
closed by #63 ?
